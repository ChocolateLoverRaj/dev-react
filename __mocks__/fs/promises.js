/* global jest */

// Mocker for fs/promises

const fs = jest.createMockFromModule('fs/promises')

const EventEmitter = require('eventemitter3')

const frozen = new Set()
fs._frozen = frozen
const errorFiles = new Set()
fs._errorFiles = errorFiles
const files = new Map()
fs._files = files
fs._reset = () => {
  frozen.clear()
  errorFiles.clear()
  files.clear()
}

const mock = new EventEmitter()
  .on('freeze', filename => {
    frozen.add(filename)
  })
  .on('unfreeze', filename => {
    frozen.delete(filename)
  })
  .on('throw', filename => {
    errorFiles.add(filename)
  })
  .on('reset', () => {
    fs._reset()
  })
fs._mock = mock

const onceUnfrozen = filename => new Promise((resolve, reject) => {
  const handler = f => {
    if (f === filename) {
      mock.off(handler)
      resolve()
    }
  }
  mock.on('unfreeze', handler)
})

fs.writeFile = async (filename, content) => {
  if (errorFiles.has(filename)) {
    throw new Error('Error writing file.')
  }
  const write = () => {
    files.set(filename, content)
    mock.emit('wrote', filename, content)
  }
  if (frozen.has(filename)) {
    await onceUnfrozen(filename)
    write()
  } else {
    write()
  }
}

fs.readFile = async filename => {
  if (errorFiles.has(filename)) {
    throw new Error('Error reading file.')
  }
  const read = () => {
    mock.emit('read', filename)
    return files.get(filename)
  }
  if (frozen.has(filename)) {
    await onceUnfrozen(filename)
    return read()
  } else {
    return read()
  }
}

fs.unlink = async filename => {
  files.delete(filename)
  mock.emit('unlinked', filename)
}

module.exports = fs
