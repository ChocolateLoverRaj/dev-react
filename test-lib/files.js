export class File {
  constructor () {
    this.canRead = true
    this.canWrite = true
  }
}

export class NormalFile extends File {
  constructor (content) {
    super()
    this.content = content
  }
}

export class Dir extends File {
  constructor () {
    super()
    this.files = new Map()
  }
}

class EnoentError extends Error {
  constructor () {
    super(...arguments)
    this.code = 'ENOENT'
  }
}

export const topDir = new Dir()

export const reset = () => {
  topDir.files.clear()
}

export const getFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
  return file
}

export const setFile = (path, fileToSet) => {
  let dir = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (fileTree.length === filenames.length) {
      dir.files.set(filename, fileToSet)
      return
    }
    if (dir.files.has(filename)) {
      dir = dir.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}

export const unlinkFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      if (fileTree.length === filenames.length) {
        file.files.delete(filename)
        return
      }
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}
