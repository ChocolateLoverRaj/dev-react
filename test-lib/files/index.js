// TODO: structure with folders and files
const topDir = new Map()

export default topDir