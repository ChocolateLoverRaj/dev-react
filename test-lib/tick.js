// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

module.exports = tick
