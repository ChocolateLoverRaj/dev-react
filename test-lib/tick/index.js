// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

export default tick
