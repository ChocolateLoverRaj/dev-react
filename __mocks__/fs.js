// Mocker for fs
import { getFile } from '../test-lib/files.js'
import tick from '../test-lib/tick.js'
import { Readable, PassThrough } from 'stream'

export const _readStreamErrorFiles = new Set()

export const _reset = () => {
  _readStreamErrorFiles.clear()
}

export const createReadStream = path => {
  if (_readStreamErrorFiles.has(path)) {
    const stream = new PassThrough()
    tick().then(() => {
      stream.destroy(new Error('Error creating a read stream.'))
    })
    return stream
  }
  return Readable.from(getFile(path).content.split(''))
}
