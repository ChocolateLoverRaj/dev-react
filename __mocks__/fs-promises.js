// Mocker for fs/promises
import { NormalFile, reset, unlinkFile, setFile, getFile } from '../test-lib/files.js'
import { constants } from './fs.js'
import EventEmitter from 'eventemitter3'

export const _frozen = new Set()
export const _errorFiles = new Set()
export const _reset = () => {
  _frozen.clear()
  _errorFiles.clear()
  reset()
}

export const _mock = new EventEmitter()
  .on('unfreeze', filename => {
    _frozen.delete(filename)
  })

const onceUnfrozen = filename => new Promise(resolve => {
  const handler = f => {
    if (f === filename) {
      _mock.off('unfreeze', handler)
      resolve()
    }
  }
  _mock.on('unfreeze', handler)
})

export const access = async (filename, permissions) => {
  const file = getFile(filename)
  const filePermissions = file.canRead * constants.R_OK | file.canWrite * constants.W_OK
  if (permissions !== permissions & filePermissions) {
    throw new Error('Bad permissions.')
  }
}

export const writeFile = async (filename, content) => {
  if (_errorFiles.has(filename)) {
    throw new Error('Error writing file.')
  }
  const write = () => {
    setFile(filename, new NormalFile(content))
    _mock.emit('wrote', filename, content)
  }
  if (_frozen.has(filename)) {
    await onceUnfrozen(filename)
    write()
  } else {
    write()
  }
}

export const unlink = async filename => {
  if (_errorFiles.has(filename)) {
    throw new Error('Error unlinking file.')
  }
  unlinkFile(filename)
}
