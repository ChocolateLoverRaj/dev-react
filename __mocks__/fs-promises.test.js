/* global afterEach, test, expect, describe */

import { access, writeFile, unlink, _frozen, _errorFiles, _reset, _mock } from './fs-promises.js'
import { reset, getFile, setFile, NormalFile } from '../test-lib/files.js'
import tick from '../test-lib/tick.js'
import noResolve from '../test-lib/no-resolve.js'
import { constants } from './fs'

afterEach(() => {
  reset()
})
afterEach(_reset)

describe('access', () => {
  test('ENOENT', async () => {
    await expect(access('file')).rejects.toMatchSnapshot()
  })

  test('Read Only', async () => {
    const file = new NormalFile('hi')
    file.canWrite = false
    setFile('file', file)
    await access('file', constants.R_OK)
  })

  test('Read and Write', async () => {
    const file = new NormalFile('hi')
    file.canWrite = false
    setFile('file', file)
    await expect(access(file, constants.R_OK | constants.W_OK)).rejects.toMatchSnapshot()
  })
})

test('writeFile', async () => {
  await writeFile('file', 'data')
  expect(getFile('file').content).toBe('data')
})

test('unlink', async () => {
  setFile('file', new NormalFile('something'))
  await unlink('file')
  expect(() => { getFile('file') }).toThrowError()
})

test('freezing', async () => {
  _frozen.add('file')
  const freezePromise = (async () => {
    await tick()
    _mock.emit('unfreeze', 'file')
  })()
  await noResolve(writeFile('file', 'data'), freezePromise)
  expect(getFile('file').content).toBe('data')
})

describe('_errorFiles', () => {
  test('writeFile', async () => {
    _errorFiles.add('file')
    await expect(writeFile('file')).rejects.toMatchSnapshot()
  })

  test('unlink', async () => {
    _errorFiles.add('file')
    await expect(unlink('file')).rejects.toMatchSnapshot()
  })
})

test('_reset', () => {
  _errorFiles.add('file')
  _reset()
  expect(_errorFiles.size).toBe(0)
})

test('ENOENT', async () => {
  let threw = false
  try {
    await unlink('file')
  } catch (e) {
    threw = true
    expect(e.code).toBe('ENOENT')
    expect(e).toMatchSnapshot()
  }
  if (!threw) {
    throw new Error('Unlink did not throw')
  }
})
