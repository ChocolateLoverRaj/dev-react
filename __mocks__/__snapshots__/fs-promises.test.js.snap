// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_errorFiles unlink 1`] = `[Error: Error unlinking file.]`;

exports[`_errorFiles writeFile 1`] = `[Error: Error writing file.]`;

exports[`ENOENT 1`] = `[Error: File: file does not exist]`;

exports[`access ENOENT 1`] = `[Error: File: file does not exist]`;

exports[`access Read and Write 1`] = `[TypeError: path.split is not a function]`;
